File: arrayDivision.java
import java.util.*;

class Solution {
    public boolean isValid(List<Integer> a, int k, int mid) {
        int groups = 0;
        int sum = 0;
        int n = a.size();

        for (int i = 0; i < n; i++) {
            if (a.get(i) > mid)
                return false;

            if (sum + a.get(i) > mid) {
                groups++;
                sum = 0;
            }
            sum += a.get(i);
        }

        if (sum > 0)
            groups++;

        return groups <= k;
    }

    public int arrayDivision(List<Integer> a, int k) {
        int lo = 0;
        int hi = 0;
        int ans = 0;

        for (int num : a)
            hi += num;

        while (lo <= hi) {
            int mid = (hi + lo) / 2;

            if (isValid(a, k, mid)) {
                hi = mid - 1;
                ans = mid;
            } else {
                lo = mid + 1;
            }
        }

        return ans;
    }
}

public class Main {
    public static void main(String[] args) {
        Solution obj = new Solution();

        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int k = scanner.nextInt();

        List<Integer> a = new ArrayList<>();

        for (int i = 0; i < n; i++)
            a.add(scanner.nextInt());

        System.out.println(obj.arrayDivision(a, k));
    }
}

File: arrayDivision.py
class Solution:
    def isValid(self, a, k, mid):
        groups = 0
        summ = 0
        n = len(a)

        for i in range(n):
            if a[i] > mid:
                return False

            if summ + a[i] > mid:
                groups += 1
                summ = 0
            summ += a[i]

        if summ > 0:
            groups += 1

        return groups <= k

    def arrayDivision(self, a, k):
        lo = 0
        hi = 0
        ans = 0

        for num in a:
            hi += num

        while lo <= hi:
            mid = (hi + lo) // 2

            if self.isValid(a, k, mid):
                hi = mid - 1
                ans = mid
            else:
                lo = mid + 1

        return ans


obj = Solution()

n, k = map(int, input().split())
a = list(map(int, input().split()))

print(obj.arrayDivision(a, k))
